### PR TITLE
92 avro single message read

### PR DESCRIPTION
**Pull request by wojtkiewicz**

### CODE REVIEW

**Comment by druminski** [comment]
:+1:

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java [hunk 1]
```diff
@@ -1,72 +1,85 @@
...
-        for (MessageAndOffset messageAndOffset : fetchResponse.messageSet(topicName, partition)) {
-            long currentOffset = messageAndOffset.offset();
-            if (currentOffset < offset) {
-                logger.info("Found an old offset: {} Expecting: {}", currentOffset, offset);
-                continue;
-            }
-            ByteBuffer payload = messageAndOffset.message().payload();
+    private byte[] convertAvroToJson(String schema, byte[] bytes) {
+        return new AvroToJsonConverter(new Schema.Parser().parse(schema))
+                .convert(avroMessageContentWrapper.unwrapContent(bytes).getContent());
```
Comment (druminski) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java]: Unwrapped content is returned for avro message, for json is wrapped. Maybe we should somehow unify it?

Comment (wojtkiewicz) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java]: Good question. Maybe always returning wrapped messages is a right call? Right now I cannot convert AVRO to JSON without some AVRO schema containing metadata though...

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java [hunk 2]
```diff
@@ -23,32 +28,34 @@ public UnwrappedMessageContent unwrapContent(byte[] data) {
...
-            int wrappedMessageSize = metadataSize + message.length;
-
-            byte [] wrappedMessage = new byte[wrappedMessageSize];
-            System.arraycopy(metadataBuffer, 0, wrappedMessage, 0, metadataSize);
-            System.arraycopy(message, 0, wrappedMessage, metadataSize, message.length);
-
-            return wrappedMessage;
-
-        } catch (RuntimeException exception) {
+            ByteArrayOutputStream wrappedMessage = new ByteArrayOutputStream();
```
Comment (pjagielski) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java]: :heart:

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java [hunk 3]
```diff
@@ -1,72 +1,83 @@
...
-            createFetchRequest(simpleConsumer.clientId(), topic, partition, offset, simpleConsumerPool.getBufferSize())
-        );
-
-        String topicName = topic.qualifiedName();
-        if (fetchResponse.hasError()) {
-            logger.error("Cannot read offset {} from topic/partition {}/{}. Error code: {}",
-                    offset, topic.qualifiedName(), partition, fetchResponse.errorCode(topicName, partition));
-            throw new SingleMessageReaderException(String.format("Cannot read offset %d from topic/partition %s/%d. Error code: %d",
-                    offset, topic.qualifiedName(), partition, fetchResponse.errorCode(topicName, partition)));
+    public String readMessage(Topic topic, int partition, long offset) {
```
Comment (pjagielski) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java]: maybe we can just return bytes here and do conversion later?

**Comment by druminski** [comment]
you can remove hermesMessage.avsc file because it can misleads now. Besides few small comments LGTM

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java [hunk 4]
```diff
@@ -23,32 +28,34 @@ public UnwrappedMessageContent unwrapContent(byte[] data) {
...
+    public Schema getWrappedSchema(Schema messageSchema) {
+        return record("MessageWithMetadata")
+                .namespace("pl.allegro.tech.hermes")
+                .fields()
+                    .name("metadata").type().record("MessageMetadata")
+                        .fields()
+                            .name("timestamp").type().longType().noDefault()
+                            .name("id").type().stringType().noDefault()
+                        .endRecord().noDefault()
+                    .name("message").type(messageSchema).noDefault()
```
Comment (druminski) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java]: this particular line is pretty cool

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java (hunk 2 above)
Comment (adamdubiel) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java]: <3

**Comment by adamdubiel** [comment]
LGTM :+1:

File: hermes-common/src/test/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapperTest.java [hunk 5]
```diff
@@ -4,32 +4,33 @@
...
+
+    private final String id = UUID.randomUUID().toString();
+    private final Long timestamp = System.currentTimeMillis();
 
     @Before
     public void setup() throws IOException {
+        avroUser = new AvroUser();
+        content = avroUser.create("Bob", 10, "red");
         avroMessageContentWrapper = new AvroMessageContentWrapper();
-        schema = new Schema.Parser().parse(getClass().getResourceAsStream("/hermesMessage.avsc"));
```
Comment (pawelszymczyk) [inline hermes-common/src/test/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapperTest.java]: hermesMessage.avsc can be deleted from resources?

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java (hunk 3 above)
Comment (wojtkiewicz) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java]: done

File: hermes-common/src/test/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapperTest.java (hunk 5 above)
Comment (wojtkiewicz) [inline hermes-common/src/test/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapperTest.java]: done

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java (hunk 4 above)
Comment (wojtkiewicz) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageContentWrapper.java]: :sunglasses:

**Comment by pjagielski** [comment]
LGBT :+1:

**Comment by pawelszymczyk** [comment]
:+1:

**Pull request merged**